### PR TITLE
DEPR: astype dt64<->dt64tz

### LINKED ISSUE #38622: BUG: dt64 <-> dt64tz roundup
xref #33401 which discusses astype, #24559 which discusses constructors.  I think these need to be considered jointly.

Two big-picture questions.

1) when we pass `constructor(dt64values, tz=tz)`, do we interpret dt64values as UTC or wall-times?
2) when we pass constructor(dt64values).astype("M8[ns, tz]")`, do we interpret dt64values as UTC or wall-times?

Let's check the current behavior on master for Timestamp, DatetimeIndex, and Series (see code block below if you want to check my work)

 cls                                        | Timestamp  | DatetimeIndex | Series  |
------------------------ | ------------- | ------------- | ------- |
cls(dt64, dtype=dt64tz)  | UTC  | Wall  | UTC |
cls(dt64).astype(dt64tz) | N/A  | Wall  | UTC |
cls(dt64tz).astype("M8[ns]")  | N/A  | UTC | UTC |

If we were starting fresh, I would probably lean towards making these all interpret dt64 as wall-times.  But because these are already mostly-UTC, it would be an easier change to bring the DatetimeIndex behavior in line with the others.

Thoughts?  cc @jorisvandenbossche @mroeschke @jreback 

```
values = (3600 * 10**9 * np.arange(4)).astype("datetime64[ns]")
dt64 = values[-1]

tz = "US/Pacific"
dtype = pd.DatetimeTZDtype(tz=tz)

ts = pd.Timestamp(dt64, tz=tz)
dti = pd.DatetimeIndex(values, dtype=dtype)
ser = pd.Series(values, dtype=dtype)

if_wall = pd.DatetimeIndex(values).tz_localize(tz)
if_utc = pd.DatetimeIndex(values).tz_localize("UTC").tz_convert(tz)

assert ts == if_utc[-1]
assert (dti == if_wall).all()
assert (ser == if_utc).all()

dti_naive = pd.DatetimeIndex(values)
ser_naive = pd.Series(values)

res_dti = dti_naive.astype(dtype)
res_ser = ser.astype(dtype)

assert (res_dti == dti).all()
assert (res_ser == ser).all()

# astype from tzaware -> tznaive
assert (dti.astype(dti_naive.dtype) == dti.tz_convert("UTC").tz_localize(None)).all()
assert (ser_naive.astype(ser.dtype) == ser).all()
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #38622
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/user_guide/timeseries.rst [hunk 1]
```diff
@@ -2605,17 +2605,10 @@ For example, to localize and convert a naive stamp to time zone aware.
...
+This method can convert between different timezone-aware dtypes.
 
 .. ipython:: python
 
-   # localize and convert a naive time zone
-   s_naive.astype("datetime64[ns, US/Eastern]")
-
-   # make an aware tz naive
-   s_aware.astype("datetime64[ns]")
-
```
Comment (jreback) [inline doc/source/user_guide/timeseries.rst:2618]: we are still allowing essentially a convert between 2 timezones right?

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -942,6 +942,32 @@ def coerce_indexer_dtype(indexer, categories):
     return ensure_int64(indexer)
 
 
+def _find_level() -> int:
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:945]: -> _find_stack_level

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -942,6 +942,32 @@ def coerce_indexer_dtype(indexer, categories):
...
+    """
+    Find the appropriate stacklevel with which to issue a warning for astype.
+    """
+    import inspect
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:949]: can import at the top

File: pandas/core/dtypes/cast.py [hunk 4]
```diff
@@ -942,6 +942,32 @@ def coerce_indexer_dtype(indexer, categories):
     return ensure_int64(indexer)
 
 
+def _find_level() -> int:
+    """
+    Find the appropriate stacklevel with which to issue a warning for astype.
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:947]: i would move this elsewhere, maybe to pandas.util._validation (or similar).

maybe just pass the names in as needed e.g. L958 (or hard-code ok too)

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
tested with DTI, Series, DTA? (i think so, just confirming). do we allow this with Timestamp? (obviously deprecate if we do)

File: doc/source/user_guide/timeseries.rst (hunk 1 above)
Comment (jbrockmendel) [reply doc/source/user_guide/timeseries.rst:2618]: yes, e.g. the example on L2620 below

File: pandas/core/dtypes/cast.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:947]: maybe util._exceptions?

Comment (jreback) [reply pandas/core/dtypes/cast.py:947]: sure

**Comment by jbrockmendel** [comment]
> tested with DTI, Series, DTA? (i think so, just confirming). 

Yes

> do we allow this with Timestamp? (obviously deprecate if we do)

No, there is no Timestamp.astype

File: pandas/core/dtypes/cast.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py]: updated + green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
great

**Pull request merged**